File: holoken-core/src/test/java/com/holokenmod/creation/TestGridCalculator.java
package com.holokenmod.creation;

import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.MatcherAssert.assertThat;

import com.holokenmod.backtrack.hybrid.MathDokuCage2BackTrack;
import com.holokenmod.grid.Grid;
import com.holokenmod.grid.GridCage;
import com.holokenmod.grid.GridCell;
import com.holokenmod.grid.GridSize;
import com.holokenmod.options.CurrentGameOptionsVariant;
import com.holokenmod.options.DigitSetting;
import com.holokenmod.options.GridCageOperation;
import com.holokenmod.options.SingleCageUsage;

import org.junit.jupiter.api.RepeatedTest;

public class TestGridCalculator {
	@RepeatedTest(1)
	void test3x3Grid() {
		CurrentGameOptionsVariant.getInstance().setDigitSetting(DigitSetting.FIRST_DIGIT_ONE);
		CurrentGameOptionsVariant.getInstance().setShowOperators(true);
		CurrentGameOptionsVariant.getInstance().setSingleCageUsage(SingleCageUsage.FIXED_NUMBER);
		CurrentGameOptionsVariant.getInstance().setCageOperation(GridCageOperation.OPERATIONS_ALL);
		
		GridCalculator creator = new GridCalculator(
				new GridSize(9, 9));
		
		Grid grid = creator.calculate();
		
		MathDokuCage2BackTrack backTrack = new MathDokuCage2BackTrack(grid, false);
		
		int solutions = backTrack.solve();
		
		assertThat(
				"Found " + solutions + " solutions, but there should be exactly one. " + grid,
				solutions,
				is(1));
	}
	
	@RepeatedTest(1)
	void bruteForce() {
		CurrentGameOptionsVariant.getInstance().setDigitSetting(DigitSetting.FIRST_DIGIT_ONE);
		CurrentGameOptionsVariant.getInstance().setShowOperators(true);
		CurrentGameOptionsVariant.getInstance().setSingleCageUsage(SingleCageUsage.FIXED_NUMBER);
		CurrentGameOptionsVariant.getInstance().setCageOperation(GridCageOperation.OPERATIONS_ALL);
		
		GridCalculator creator = new GridCalculator(
				new GridSize(4, 4));
		
		
		Grid grid = creator.calculate();
		
		solveBruteForce(grid, 0);
	}
	
	private void solveBruteForce(Grid grid, int cellNumber) {
		if (cellNumber == grid.getGridSize().getSurfaceArea()) {
			
			if (isValidSolution(grid)) {
				System.out.println("Found valid solution.");
				
				for(GridCell cell: grid.getCells()) {
					assertThat(
							"Found differing solution. " + grid,
							cell.getUserValue(), is(cell.getValue()));
				}
			}
			
			return;
		}
		
		GridCell cell = grid.getCell(cellNumber);

		for(int value : grid.getPossibleDigits()) {
			if (!grid.isUserValueUsedInSameColumn(cellNumber, value)
					&& !grid.isUserValueUsedInSameRow(cellNumber, value)) {
				
				cell.setUserValueIntern(value);
				
				solveBruteForce(grid, cellNumber + 1);
			}
		}
		
		cell.setUserValueIntern(GridCell.NO_VALUE_SET);
	}
	
	private boolean isValidSolution(Grid grid) {
		boolean validSolution = true;
		
		for( GridCell cell: grid.getCells()) {
			validSolution &= !grid.isUserValueUsedInSameColumn(cell.getCellNumber(), cell.getUserValue());
			validSolution &= !grid.isUserValueUsedInSameRow(cell.getCellNumber(), cell.getUserValue());
		}
		
		for(GridCage cage : grid.getCages()) {
			validSolution &= cage.isMathsCorrect();
		}
		
		return validSolution;
	}
}

File: holoken-core/src/test/java/com/holokenmod/creation/TestGridCalculator.kt
package com.holokenmod.creation

import com.holokenmod.backtrack.hybrid.MathDokuCage2BackTrack
import com.holokenmod.grid.*
import com.holokenmod.options.GameOptionsVariant.Companion.createClassic
import com.holokenmod.options.GameVariant
import org.hamcrest.CoreMatchers
import org.hamcrest.MatcherAssert
import org.junit.jupiter.api.RepeatedTest

class TestGridCalculator {
    @RepeatedTest(1)
    fun test3x3Grid() {
        val creator = GridCalculator(
            GameVariant(
                GridSize(9, 9),
                createClassic()
            )
        )
        val grid = creator.calculate()
        val backTrack = MathDokuCage2BackTrack(grid, false)
        val solutions = backTrack.solve()
        MatcherAssert.assertThat(
            "Found $solutions solutions, but there should be exactly one. $grid",
            solutions,
            CoreMatchers.`is`(1)
        )
    }

    @RepeatedTest(1)
    fun bruteForce() {
        val creator = GridCalculator(
            GameVariant(
                GridSize(4, 4),
                createClassic()
            )
        )
        val grid = creator.calculate()
        solveBruteForce(grid, 0)
    }

    private fun solveBruteForce(grid: Grid, cellNumber: Int) {
        if (cellNumber == grid.gridSize.surfaceArea) {
            if (isValidSolution(grid)) {
                println("Found valid solution.")
                for (cell in grid.cells) {
                    MatcherAssert.assertThat(
                        "Found differing solution. $grid",
                        cell.userValue, CoreMatchers.`is`(cell.value)
                    )
                }
            }
            return
        }
        val cell = grid.getCell(cellNumber)
        for (value in grid.possibleDigits) {
            if (!grid.isUserValueUsedInSameColumn(cellNumber, value)
                && !grid.isUserValueUsedInSameRow(cellNumber, value)
            ) {
                cell.setUserValueIntern(value)
                solveBruteForce(grid, cellNumber + 1)
            }
        }
        cell.setUserValueIntern(GridCell.NO_VALUE_SET)
    }

    private fun isValidSolution(grid: Grid): Boolean {
        var validSolution = true
        for (cell in grid.cells) {
            validSolution =
                validSolution and !grid.isUserValueUsedInSameColumn(cell.cellNumber, cell.userValue)
            validSolution =
                validSolution and !grid.isUserValueUsedInSameRow(cell.cellNumber, cell.userValue)
        }
        for (cage in grid.cages) {
            validSolution = validSolution and cage.isMathsCorrect()
        }
        return validSolution
    }
}